File: framework/languages/java/e2e/ResilienceLifecycle/Client/src/main/java/systems/zlink/e2e/resiliencelifecycle/client/Scenarios/RlA4DrainAndGreenEndpointScenario.java
package systems.zlink.e2e.resiliencelifecycle.client.Scenarios;

import systems.zlink.e2e.resiliencelifecycle.client.Support.ResilienceScenarioContext;

public final class RlA4DrainAndGreenEndpointScenario {
    private RlA4DrainAndGreenEndpointScenario() {
    }

    public static void run(ResilienceScenarioContext context) {
        context.waitForTopology(2);
        context.post(context.adminB() + "/admin/drain");
        context.waitForWeight(context.adminB(), 0);
        context.collectStableProvidersWithout("a4-before-green", "api-b", "api-a");
        context.signal("a4-drained");
        context.waitForSignal("a4-original-down");
        context.collectStableProvidersWithout("a4-old-down", "api-b", "api-a");
        context.waitForSignal("a4-green-up");
        context.waitForTopologyEndpoint(
            "api-b", context.options().apiBGreenEndpoint());
        context.driveUntilEvidence(
            context.adminBGreen(), "a4-green", "RL-A4 green provider did not receive traffic");
        context.signal("a4-green-observed");
        context.signal("a4-restore-ready");
        context.waitForSignal("a4-restored");
        context.waitForTopologyEndpoint("api-b", context.options().apiBEndpoint());
        context.driveUntilEvidence(
            context.adminB(), "a4-restored", "RL-A4 restored provider did not receive traffic");
        System.out.println("scenario RL-A4 passed");
    }
}

File: framework/languages/java/e2e-kotlin/ResilienceLifecycle/Client/src/main/kotlin/systems/zlink/e2e/kotlin/resiliencelifecycle/Scenarios/RlA4DrainAndGreenEndpointScenario.kt
package systems.zlink.e2e.kotlin.resiliencelifecycle

fun ClientScenarioContext.runDrainAndGreenEndpointScenario() {
    val greenApi = options.apiBGreenEndpoint
        ?: throw IllegalStateException("e2e.api.b.green.endpoint is required")
    val greenHttp = options.httpBGreenEndpoint
        ?: throw IllegalStateException("e2e.http.b.green.endpoint is required")

    waitForTopology(2)
    post("${adminB()}/admin/drain")
    waitForWeight(adminB(), 0)
    collectStableProvidersWithout("a4-rolling", "api-b", "api-a")

    signal("a4-drained")
    waitForSignal("a4-green-up")
    waitForTopologyEndpoint("api-b", greenApi)
    driveProviderTraffic("a4-green", greenHttp)

    signal("a4-green-served")
    waitForSignal("a4-green-down")
    waitForTopologyMissing("api-b")

    signal("a4-restore")
    waitForSignal("a4-restored")
    waitForTopologyEndpoint(
        "api-b",
        options.apiBEndpoint ?: throw IllegalStateException("e2e.api.b.endpoint is required"),
    )
    driveProviderTraffic("a4-restored", adminB())
    println("scenario RL-A4 passed")
}

private fun ClientScenarioContext.driveProviderTraffic(prefix: String, providerHttp: String) {
    repeat(120) { index ->
        val marker = "$prefix-$index"
        val reply = requestWork(marker)
        ensure(reply.value() == "work:$marker", "RL-A4 $prefix payload mismatch")
        if (reply.providerRid() == "api-b") {
            waitForEvidenceValue(providerHttp, "WorkReq", marker)
            return
        }
        sleep(100)
    }
    throw IllegalStateException("RL-A4 $prefix traffic did not reach api-b")
}
